### PR TITLE
Optimize babel-template

### PR DESCRIPTION
**Pull request by jridgewell**
| Q                 | A 
| ----------------- | ---
| Bug fix?          | No
| Breaking change?  | No
| New feature?      | No
| Deprecations?     | No
| Spec compliancy?  | No
| Tests added/pass? | No
| Fixed tickets     | 
| License           | MIT
| Doc PR            | 
| Dependency Changes| 

Only enter into `Identifier`s, avoid modifying the nodes, skip to avoid entering into replaced nodes.

### CODE REVIEW

**Comment by hzoo** [comment]
Awesome - @jridgewell were you just looking at the source or were their perf issues you measured?

I was wondering whether we should make it more powerful and allow not just replacing identifiers but other nodes too (would be a separate discussion/pr)

**Comment by jridgewell** [comment]
Just read the code, was curious how it worked.

> but other nodes too

Do you have an example?

**Comment by hzoo** [comment]
Wonder how we can measure if were doing an optimization

Someone posted about "quasiquotes" which none of us had heard of (from functional programming) - I believe it was something like https://github.com/rektide/quasilon/blob/master/quasilon.js but I didn't really check.

**Comment by DrewML** [comment]
> Wonder how we can measure if were doing an optimization

I think it's important that we have metrics around this before merging. Since `babel-templates` is such a critical piece of the project, it wouldn't be worth making a change like this unless we knew for sure that it had a perf advantage.

**Comment by loganfsmyth** [comment]
I think this helps readability of this module a ton and I don't see anything I'd expect to be a perf issue.

File: packages/babel-template/src/index.js [hunk 1]
```diff
@@ -77,37 +68,34 @@ let templateVisitor = {
...
+      path = parentPath;
     }
 
-    if (replacement) {
-      replacement[TEMPLATE_SKIP] = true;
+    if (replacement === null) {
+      path.remove();
+    } else if (replacement) {
       path.replaceInline(replacement);
+      path.skip();
```
Comment (loganfsmyth) [inline packages/babel-template/src/index.js]: I think this is going to fail if `replacement` is an array of nodes. Not sure what the right fix would be.

Comment (jridgewell) [inline packages/babel-template/src/index.js]: Think I've addressed. Little did I know, `path.skip()` doesn't skip `enter` either.

**Comment by loganfsmyth** [comment]
Looks like part of this might still be applicable, but the part for `babel-template` isn't right anymore.

**Comment by danez** [comment]
I tired rebasing, but seems all the changes have already been incorporated in some way, so I ended up with no changes after rebasing. Closing.

**Pull request closed without merging**